File: typescript/tests/test_ows_signer.ts
/**
 * OWS Signer unit tests.
 *
 * Uses a mock OWS module — no real @open-wallet-standard/core needed.
 * Tests construction, EIP-712 typed data building, signature concatenation,
 * BigInt serialization, error paths, and security (no key leakage).
 */

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { OwsSigner } from "../src/ows-signer.js";

// ── Mock OWS module ──────────────────────────────────────────────────

interface SignCall {
  wallet: string;
  chain: string;
  json: string;
  passphrase?: string;
}

function createMockOws(options?: {
  accounts?: Array<{ chainId: string; address: string; derivationPath: string }>;
  signature?: string;
  recoveryId?: number;
}) {
  const calls: SignCall[] = [];
  const accounts = options?.accounts ?? [
    {
      chainId: "eip155:8453",
      address: "0x1234567890abcdef1234567890abcdef12345678",
      derivationPath: "m/44'/60'/0'/0/0",
    },
  ];

  return {
    calls,
    getWallet(nameOrId: string) {
      return {
        id: `id-${nameOrId}`,
        name: nameOrId,
        accounts,
        createdAt: "2026-04-01T00:00:00Z",
      };
    },
    signTypedData(
      wallet: string,
      chain: string,
      typedDataJson: string,
      passphrase?: string,
    ) {
      calls.push({ wallet, chain, json: typedDataJson, passphrase });
      return {
        signature: options?.signature ?? "aa".repeat(64),
        recoveryId: options?.recoveryId ?? 0,
      };
    },
  };
}

// ── Construction ─────────────────────────────────────────────────────

describe("OwsSigner.create", () => {
  it("creates signer with eip155 chain id", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });
    assert.equal(
      signer.address,
      "0x1234567890abcdef1234567890abcdef12345678",
    );
  });

  it("creates signer with evm chain id", async () => {
    const mock = createMockOws({
      accounts: [
        {
          chainId: "evm",
          address: "0xcafe",
          derivationPath: "m/44'/60'/0'/0/0",
        },
      ],
    });
    const signer = await OwsSigner.create({
      walletId: "pay-evm",
      _owsModule: mock,
    });
    assert.equal(signer.address, "0xcafe");
  });

  it("throws when no EVM account found", async () => {
    const mock = createMockOws({
      accounts: [
        {
          chainId: "solana",
          address: "Sol123",
          derivationPath: "m/44'/501'/0'/0'",
        },
      ],
    });
    await assert.rejects(
      () => OwsSigner.create({ walletId: "pay-sol", _owsModule: mock }),
      { message: /No EVM account found/ },
    );
  });

  it("throws when no accounts at all", async () => {
    const mock = createMockOws({ accounts: [] });
    await assert.rejects(
      () => OwsSigner.create({ walletId: "pay-empty", _owsModule: mock }),
      { message: /No EVM account found/ },
    );
  });

  it("throws clear error when OWS not installed", async () => {
    // Without _owsModule, create() will try dynamic import which will fail
    await assert.rejects(
      () => OwsSigner.create({ walletId: "pay-missing" }),
      { message: /not installed/ },
    );
  });
});

// ── signTypedData ────────────────────────────────────────────────────

describe("OwsSigner.signTypedData", () => {
  const domain = {
    name: "Pay",
    version: "1",
    chainId: 8453,
    verifyingContract: "0xrouter" as const,
  };

  const types = {
    Request: [
      { name: "method", type: "string" },
      { name: "path", type: "string" },
    ],
  };

  const value = { method: "POST", path: "/api/v1/direct" };

  it("builds EIP-712 JSON with EIP712Domain injected", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    await signer.signTypedData(domain, types, value);

    assert.equal(mock.calls.length, 1);
    const parsed = JSON.parse(mock.calls[0].json);

    // EIP712Domain should be auto-generated from domain fields
    assert.deepStrictEqual(parsed.types.EIP712Domain, [
      { name: "name", type: "string" },
      { name: "version", type: "string" },
      { name: "chainId", type: "uint256" },
      { name: "verifyingContract", type: "address" },
    ]);

    // Original types preserved
    assert.deepStrictEqual(parsed.types.Request, types.Request);

    // Primary type derived from first non-EIP712Domain key
    assert.equal(parsed.primaryType, "Request");

    // Domain and message passed through
    assert.deepStrictEqual(parsed.domain, domain);
    assert.deepStrictEqual(parsed.message, value);
  });

  it("passes chain as evm always", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    await signer.signTypedData(domain, types, value);
    assert.equal(mock.calls[0].chain, "evm");
  });

  it("passes API key as passphrase", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      owsApiKey: "ows_key_secret123",
      _owsModule: mock,
    });

    await signer.signTypedData(domain, types, value);
    assert.equal(mock.calls[0].passphrase, "ows_key_secret123");

    // API key must NOT appear in the JSON payload
    assert.ok(!mock.calls[0].json.includes("ows_key_secret123"));
  });

  it("only includes domain fields that are present", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    // Domain with only name and chainId
    await signer.signTypedData(
      { name: "Pay", chainId: 8453 },
      types,
      value,
    );

    const parsed = JSON.parse(mock.calls[0].json);
    assert.deepStrictEqual(parsed.types.EIP712Domain, [
      { name: "name", type: "string" },
      { name: "chainId", type: "uint256" },
    ]);
  });

  it("serializes BigInt values to strings", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    await signer.signTypedData(domain, types, {
      method: "POST",
      path: "/direct",
      amount: 5000000n,
    });

    const parsed = JSON.parse(mock.calls[0].json);
    assert.equal(parsed.message.amount, "5000000");
  });
});

// ── Signature concatenation ──────────────────────────────────────────

describe("signature concatenation", () => {
  it("appends v=27 when recoveryId=0", async () => {
    const mock = createMockOws({
      signature: "aa".repeat(64), // 128 hex = r+s only
      recoveryId: 0,
    });
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    const sig = await signer.signTypedData({}, { R: [] }, {});
    assert.ok(sig.startsWith("0x"));
    assert.equal(sig.length, 2 + 130); // 0x + 64r + 64s + 2v
    assert.equal(sig.slice(-2), "1b"); // 27 = 0x1b
  });

  it("appends v=28 when recoveryId=1", async () => {
    const mock = createMockOws({
      signature: "bb".repeat(64),
      recoveryId: 1,
    });
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    const sig = await signer.signTypedData({}, { R: [] }, {});
    assert.equal(sig.slice(-2), "1c"); // 28 = 0x1c
  });

  it("passes through 130-char RSV signature as-is", async () => {
    const rsv = "cc".repeat(65); // 130 hex = already has v
    const mock = createMockOws({ signature: rsv });
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    const sig = await signer.signTypedData({}, { R: [] }, {});
    assert.equal(sig, `0x${rsv}`);
  });

  it("strips 0x prefix from OWS signature", async () => {
    const mock = createMockOws({
      signature: "0x" + "dd".repeat(64),
      recoveryId: 0,
    });
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    const sig = await signer.signTypedData({}, { R: [] }, {});
    assert.ok(!sig.includes("0x0x")); // no double prefix
    assert.equal(sig.length, 2 + 130);
  });

  it("defaults recoveryId to 0 when undefined", async () => {
    const mock = createMockOws({
      signature: "ee".repeat(64),
      recoveryId: undefined,
    });
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    const sig = await signer.signTypedData({}, { R: [] }, {});
    assert.equal(sig.slice(-2), "1b"); // v=27
  });
});

// ── sign() rejection ─────────────────────────────────────────────────

describe("sign() raw hash rejection", () => {
  it("throws because OWS only supports EIP-712", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      _owsModule: mock,
    });

    assert.throws(
      () => signer.sign(new Uint8Array(32)),
      { message: /does not support raw hash signing/ },
    );
  });
});

// ── Security ─────────────────────────────────────────────────────────

describe("security", () => {
  it("toJSON does not expose API key", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      owsApiKey: "ows_key_topsecret",
      _owsModule: mock,
    });

    const json = signer.toJSON();
    assert.ok(!JSON.stringify(json).includes("topsecret"));
    assert.equal(json.walletId, "pay-test");
    assert.ok(json.address);
  });

  it("inspect does not expose API key", async () => {
    const mock = createMockOws();
    const signer = await OwsSigner.create({
      walletId: "pay-test",
      owsApiKey: "ows_key_topsecret",
      _owsModule: mock,
    });

    const inspectFn = signer[Symbol.for("nodejs.util.inspect.custom")] as () => string;
    const output = inspectFn.call(signer);
    assert.ok(!output.includes("topsecret"));
    assert.ok(output.includes("pay-test"));
  });
});

File: python/tests/test_ows_signer.py
"""OWS Signer unit tests.

Uses a mock OWS module — no real open-wallet-standard needed.
Tests construction, EIP-712 typed data building, signature concatenation,
error paths, and security (no key leakage).
"""

from __future__ import annotations

import json

import pytest

from payskill.ows_signer import OwsSigner

# ── Mock OWS module ──────────────────────────────────────────────────


class MockOws:
    """Fake OWS module for unit testing."""

    def __init__(
        self,
        accounts: list[dict[str, str]] | None = None,
        signature: str = "aa" * 64,
        recovery_id: int = 0,
    ):
        self.accounts = (
            accounts
            if accounts is not None
            else [
                {
                    "chain_id": "eip155:8453",
                    "address": "0x1234567890abcdef1234567890abcdef12345678",
                    "derivation_path": "m/44'/60'/0'/0/0",
                }
            ]
        )
        self._signature = signature
        self._recovery_id = recovery_id
        self.calls: list[dict[str, object]] = []

    def get_wallet(self, name_or_id: str) -> dict:
        return {
            "id": f"id-{name_or_id}",
            "name": name_or_id,
            "accounts": self.accounts,
            "created_at": "2026-04-01T00:00:00Z",
        }

    def sign_typed_data(
        self,
        wallet_id: str,
        chain: str,
        typed_data_json: str,
        ows_api_key: str | None = None,
    ) -> dict:
        self.calls.append(
            {
                "wallet": wallet_id,
                "chain": chain,
                "json": typed_data_json,
                "passphrase": ows_api_key,
            }
        )
        return {
            "signature": self._signature,
            "recovery_id": self._recovery_id,
        }


# ── Construction ──────────────────────────────────────────────────────


class TestOwsSignerConstruction:
    def test_creates_with_eip155_chain_id(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)
        assert signer.address == "0x1234567890abcdef1234567890abcdef12345678"

    def test_creates_with_evm_chain_id(self):
        mock = MockOws(
            accounts=[
                {"chain_id": "evm", "address": "0xcafe", "derivation_path": "m/44'/60'/0'/0/0"}
            ]
        )
        signer = OwsSigner(wallet_id="pay-evm", _ows_module=mock)
        assert signer.address == "0xcafe"

    def test_throws_when_no_evm_account(self):
        mock = MockOws(
            accounts=[
                {"chain_id": "solana", "address": "Sol123", "derivation_path": "m/44'/501'/0'/0'"}
            ]
        )
        with pytest.raises(ValueError, match="No EVM account found"):
            OwsSigner(wallet_id="pay-sol", _ows_module=mock)

    def test_throws_when_no_accounts(self):
        mock = MockOws(accounts=[])
        with pytest.raises(ValueError, match="No EVM account found"):
            OwsSigner(wallet_id="pay-empty", _ows_module=mock)

    def test_throws_when_ows_not_installed(self):
        with pytest.raises(ImportError, match="not installed"):
            # Without _ows_module, constructor tries to import `ows`
            OwsSigner(wallet_id="pay-missing")


# ── sign_typed_data ───────────────────────────────────────────────────


class TestSignTypedData:
    DOMAIN: dict = {
        "name": "Pay",
        "version": "1",
        "chainId": 8453,
        "verifyingContract": "0xrouter",
    }
    TYPES: dict = {
        "Request": [
            {"name": "method", "type": "string"},
            {"name": "path", "type": "string"},
        ]
    }
    VALUE: dict = {"method": "POST", "path": "/api/v1/direct"}

    @pytest.fixture()
    def signer_and_mock(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)
        return signer, mock

    @pytest.mark.asyncio()
    async def test_builds_eip712_json_with_domain_type(self, signer_and_mock):
        signer, mock = signer_and_mock
        await signer.sign_typed_data(self.DOMAIN, self.TYPES, self.VALUE)

        assert len(mock.calls) == 1
        parsed = json.loads(mock.calls[0]["json"])

        # EIP712Domain auto-generated from domain fields
        assert parsed["types"]["EIP712Domain"] == [
            {"name": "name", "type": "string"},
            {"name": "version", "type": "string"},
            {"name": "chainId", "type": "uint256"},
            {"name": "verifyingContract", "type": "address"},
        ]

        # Original types preserved
        assert parsed["types"]["Request"] == self.TYPES["Request"]

        # Primary type derived from first non-EIP712Domain key
        assert parsed["primaryType"] == "Request"

        # Domain and message passed through
        assert parsed["domain"] == self.DOMAIN
        assert parsed["message"] == self.VALUE

    @pytest.mark.asyncio()
    async def test_passes_chain_as_evm(self, signer_and_mock):
        signer, mock = signer_and_mock
        await signer.sign_typed_data(self.DOMAIN, self.TYPES, self.VALUE)
        assert mock.calls[0]["chain"] == "evm"

    @pytest.mark.asyncio()
    async def test_passes_api_key_as_passphrase(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", ows_api_key="ows_key_secret123", _ows_module=mock)
        await signer.sign_typed_data(self.DOMAIN, self.TYPES, self.VALUE)

        test_key = "ows_key_secret123"  # noqa: S105
        assert mock.calls[0]["passphrase"] == test_key
        # API key must NOT appear in the JSON payload
        assert test_key not in mock.calls[0]["json"]

    @pytest.mark.asyncio()
    async def test_only_includes_present_domain_fields(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)

        # Domain with only name and chainId
        await signer.sign_typed_data({"name": "Pay", "chainId": 8453}, self.TYPES, self.VALUE)

        parsed = json.loads(mock.calls[0]["json"])
        assert parsed["types"]["EIP712Domain"] == [
            {"name": "name", "type": "string"},
            {"name": "chainId", "type": "uint256"},
        ]


# ── Signature concatenation ──────────────────────────────────────────


class TestSignatureConcatenation:
    @pytest.mark.asyncio()
    async def test_appends_v27_when_recovery_id_0(self):
        mock = MockOws(signature="aa" * 64, recovery_id=0)
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)

        sig = await signer.sign_typed_data({}, {"R": []}, {})
        assert sig.startswith("0x")
        assert len(sig) == 2 + 130  # 0x + 64r + 64s + 2v
        assert sig[-2:] == "1b"  # 27 = 0x1b

    @pytest.mark.asyncio()
    async def test_appends_v28_when_recovery_id_1(self):
        mock = MockOws(signature="bb" * 64, recovery_id=1)
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)

        sig = await signer.sign_typed_data({}, {"R": []}, {})
        assert sig[-2:] == "1c"  # 28 = 0x1c

    @pytest.mark.asyncio()
    async def test_passthrough_130_char_rsv(self):
        rsv = "cc" * 65  # 130 hex = already has v
        mock = MockOws(signature=rsv)
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)

        sig = await signer.sign_typed_data({}, {"R": []}, {})
        assert sig == f"0x{rsv}"

    @pytest.mark.asyncio()
    async def test_strips_0x_prefix(self):
        mock = MockOws(signature="0x" + "dd" * 64, recovery_id=0)
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)

        sig = await signer.sign_typed_data({}, {"R": []}, {})
        assert "0x0x" not in sig
        assert len(sig) == 2 + 130

    @pytest.mark.asyncio()
    async def test_defaults_recovery_id_to_0(self):
        mock = MockOws(signature="ee" * 64)
        # Override recovery_id to None to test default
        original_sign = mock.sign_typed_data

        def sign_no_recovery(*args, **kwargs):
            result = original_sign(*args, **kwargs)
            result["recovery_id"] = None
            return result

        mock.sign_typed_data = sign_no_recovery  # type: ignore[assignment]

        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)
        sig = await signer.sign_typed_data({}, {"R": []}, {})
        assert sig[-2:] == "1b"  # v=27


# ── sign() rejection ─────────────────────────────────────────────────


class TestSignRawHashRejection:
    def test_throws_not_implemented(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", _ows_module=mock)
        with pytest.raises(NotImplementedError, match="does not support sign"):
            signer.sign(b"\x00" * 32)


# ── Security ─────────────────────────────────────────────────────────


class TestSecurity:
    def test_repr_does_not_expose_api_key(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", ows_api_key="ows_key_topsecret", _ows_module=mock)
        output = repr(signer)
        assert "topsecret" not in output
        assert "pay-test" in output

    def test_str_does_not_expose_api_key(self):
        mock = MockOws()
        signer = OwsSigner(wallet_id="pay-test", ows_api_key="ows_key_topsecret", _ows_module=mock)
        output = str(signer)
        assert "topsecret" not in output
